### ISSUE TITLE
Git LFS invoking GIT_ASKPASS unable to execute batch file due to special characters in account name

### ISSUE THREAD

**Issue by gemal** [desktop/desktop#3752]
### Description

when even er push I get a
```
Git credentials for https://github.com/xxxx.git not found:
exit status 1
```

I tried to log out and in again and still same thing

it works from command line

### Version

**GitHub Desktop version:** 1.0.12

**OS version:** windows 10

### Steps to Reproduce

1. [First step]
1. [Second step]
1. [and so on]

**Expected behavior:** [What you expected to happen]

**Actual behavior:** [What actually happened]

**Reproduces how often:** [What percentage of the time does it reproduce?]

### Logs

```
2018-01-09T19:20:44.391Z - info: [ui] Executing unstageAll: git reset -- . (took 3.297s)
2018-01-09T19:20:45.961Z - info: [ui] Executing createCommit: git commit -F - (took 1.395s)
2018-01-09T19:20:47.603Z - info: [ui] Executing getBranches: git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(symref)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes (took 1.643s)
2018-01-09T19:20:50.196Z - info: [ui] [AppStore.withAuthenticatingUser] account found for repository: dinnerbooking.com - gemal (has token)
2018-01-09T19:20:53.805Z - info: [ui] Executing push: git -c credential.helper= push origin testsingle --set-upstream --progress (took 3.593s)
2018-01-09T19:20:53.807Z - error: [ui] `git -c credential.helper= push origin testsingle --set-upstream --progress` exited with an unexpected code: 1.
Remote "origin" does not support the LFS locking API. Consider disabling it with:
  $ git config lfs.https://github.com/gemal/dinnerbooking.com.git/info/lfs.locksverify false

Git credentials for https://github.com/gemal/dinnerbooking.com.git not found:
exit status 1
error: failed to push some refs to 'https://github.com/gemal/dinnerbooking.com.git'

2018-01-09T19:20:53.848Z - info: [main] opening in browser: https://github.com/gemal/dinnerbooking.com/pull/new/testsingle
2018-01-09T19:20:56.796Z - info: [ui] Executing getStatus: git status --untracked-files=all --branch --porcelain=2 -z (took 1.075s)
2018-01-09T19:20:57.995Z - info: [ui] Executing getBranches: git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(symref)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes (took 2.261s)
2018-01-09T19:21:06.740Z - info: [ui] Executing getBranches: git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(symref)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes (took 1.723s)
2018-01-09T19:21:12.846Z - info: [ui] Executing getBranches: git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(symref)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes (took 1.724s)
2018-01-09T19:21:35.190Z - info: [ui] Executing getBranches: git for-each-ref --format=%(refname)%00%(refname:short)%00%(upstream:short)%00%(objectname)%00%(author)%00%(parent)%00%(symref)%00%(subject)%00%(body)%00%1F refs/heads refs/remotes (took 1.736s)
```

**Comment by iAmWillShepherd** [comment]
Does your repository use LFS?

**Comment by gemal** [comment]
yes and I have LFS installed

**Comment by shiftkey** [comment]
@gemal @iAmWillShepherd I'll point out an old issue that we didn't quite get to the bottom of #2812 but rather than moving this back to there I'd like to start fresh here. We added some extra tracing into the GUI in #3360 that you can enable like this to get more information in the logs:

1. **View** | **Toggle developer tools** to open the dev tools
1. Select the **Application** tab
1. In the sidebar, expand **Local Storage** and select **file://**
1. Add a new key name `git-trace` with a value of `1`

<img width="946" src="https://user-images.githubusercontent.com/359239/34673278-18c5dcfa-f458-11e7-9830-a532d16c8ab4.png">

Could you enable that new key and try performing the push again? I thought we'd gotten to the bottom of the locking issue by updating `git-lfs` by maybe not.

Also, could you share the contents of your `.git/config` local config as well as your `~/.gitconfig` global configuration?

**Comment by shiftkey** [comment]
@gemal 

> yes and I have LFS installed

Which version of Git LFS do you have installed? Maybe the version in Desktop is clashing with this version somehow?

**Comment by gemal** [comment]
git lfs version
git-lfs/2.3.4 (GitHub; windows amd64; go 1.8.3; git d2f6752f)

**Comment by gemal** [comment]
can I send something in private? dont want to send all the log files in public

**Comment by shiftkey** [comment]
@gemal the email address in [my profile](https://github.com/shiftkey) is fine

**Comment by gemal** [comment]
I have the same setup at work where it works. Same versions
I sent you the info by email. Let me know if you didn't receive it

**Comment by shiftkey** [comment]
@gemal thanks for the log file. I don't have a clear answer currently what's happening, but I'll walk through how I understand it and see if something clicks:

```
2018-01-09T20:58:39.958Z - error: [ui] `git -c credential.helper= push origin testsingle:testsingle --progress` exited with an unexpected code: 1.
Remote "origin" does not support the LFS locking API. Consider disabling it with:
  $ git config lfs.https://github.com/gemal/dinnerbooking.com.git/info/lfs.locksverify false
```

I'm going to ignore this warning for now, as I don't think it's the real issue.

```
21:58:37.123245 run-command.c:626       trace: run_command: 'git-remote-https' 'origin' 'https://github.com/gemal/dinnerbooking.com.git'
21:58:37.714376 run-command.c:626       trace: run_command: 'C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat' 'Username for '\''https://github.com'\'': '
21:58:38.426269 run-command.c:626       trace: run_command: 'C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat' 'Password for '\''https://gemal@github.com'\'': '
```

The tracing shows that our `GIT_ASKPASS` is being detected and invoked without error. That's good.

```
21:58:39.504090 run-command.c:626       trace: run_command: 'git-lfs' 'pre-push' 'origin' 'https://github.com/gemal/dinnerbooking.com.git'
21:58:39.608376 trace git-lfs: run_command: 'git' version
21:58:39.756355 trace git-lfs: run_command: 'git' config -l
21:58:39.855625 trace git-lfs: tq: running as batched queue, batch size of 100
```
Now we're into the Git LFS hook that runs before the actual push executes.

```
21:58:39.855625 trace git-lfs: creds: filling with GIT_ASKPASS: C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat Username for "https://github.com/gemal/dinnerbooking.com.git"
Git credentials for https://github.com/gemal/dinnerbooking.com.git not found:
exit status 1
error: failed to push some refs to 'https://github.com/gemal/dinnerbooking.com.git'
```

And it fails when Git LFS invokes our `GIT_ASKPASS` script.

The `ask-pass-trampoline.bat` script is just a way for us to bounce this authentication request back into the Desktop app, and from there we have a couple of things to do.

This is the core of `app/src/ask-pass/main.ts` - it looks at the prompt and maybe writes some output. If the `GIT_ASKPASS` script doesn't write something to `stdout`, Git or Git LFS won't be able to authenticate.

```ts
const prompt = process.argv[2]
responseForPrompt(prompt).then(response => {
  if (response) {
    process.stdout.write(response)
    process.stdout.end()
  }
})
```

[`responseForPrompt`](https://github.com/desktop/desktop/blob/8099f66788d9e3fbc3fba55ad3ec05cb86b38e5d/app/src/ask-pass/ask-pass.ts#L5) uses environment variables set by Desktop to then lookup the user's credentials in the token store. It's not reading the contents of the prompt, but there's a couple of ways that this might fail:

 - a token isn't found for the user (the log file has a check for this, and it says a token exists)
 - the environment variables are somehow unset at this point (these are set [whenever we spawn Git](https://github.com/desktop/desktop/blob/c6a57140efc82fc819826d8230e6276dd2e987d9/app/src/lib/git/authentication.ts#L32-L36)) so there shouldn't be a way to do this externally).

@technoweenie any fresh thoughts on how the interop between Git LFS and Desktop might be failing, given the above tracing?

**Comment by shiftkey** [comment]
Actually, something does stand out. Let me blockquote it here for emphasis:

> 21:58:39.855625 trace git-lfs: creds: filling with GIT_ASKPASS: C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat **Username for "https://github.com/gemal/dinnerbooking.com.git"**

The [source](https://github.com/git-lfs/git-lfs/blob/b32a7ef6245810f032e13bda8008b16536d01d7a/lfsapi/creds.go#L102) for this message is:

```
// 'ucmd' will run the GIT_ASKPASS (or core.askpass) command prompting
// for a username.
ucmd := exec.Command(a.Program, a.args(fmt.Sprintf("Username for %q", u))...)
```

Note that the first argument `Username for {domain}` isn't quoted in the same way as Git:

```
trace: run_command: 'C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat' 'Username for '\''https://github.com'\'': '
```

Note that this is failing when asking for the username, not when it tries to find the password. So my theory here is:

 - the `Username for` and `Password for` arguments passed to the batch file from `git-lfs` are not passed through to Desktop in the expected format
 - which means our `GIT_ASKPASS` implementation doesn't emit any output
 - which means `git-lfs` considers it an authentication error

Given how problematic quoting correctly inside batch files is, and how the batch file spawns a Node process which then looks at arguments, I need to think about how to test this with fewer moving parts and confirm this is the issue at play.

**Comment by shiftkey** [comment]
@gemal I'm no closer to understanding the root cause here, so when you have a moment I'd like for you to try out a test build of Desktop. It's based on the work in #3759 to add tracing to the internals of our `GIT_ASKPASS` implementation.

I'd like for you to run through these steps for me when you have a moment:

 - uninstall your version of Desktop (the data will be preserved)
 - install [this build of Desktop](https://s3.amazonaws.com/github-desktop/releases/1.0.12-test2-bd8f487b/GitHubDesktopSetup.exe)
 - run through the steps to push changes with the LFS repository so that it triggers the error
 - open `%APPDATA%\GitHub Desktop\logs\askpass.log` to see the tracing of the `GIT_ASKPASS` internals
 - review the log to ensure no sensitive data is present - I've been careful to ensure we only log what we need to understand the behaviour, but feel free to `x` out anything you're nervous about
 - email the `askpass.log` file over to me to dig into further

Thanks for working through this with me!

**Comment by shiftkey** [comment]
The awesome @gemal got back to me with the generated `askpass.log` file, and I didn't see any entries containing `"Username for \"https://github.com/owner/repo.git\""` (Git LFS adds the whole repository path when invoking `GIT_ASKPASS`). I got a bit sad at this point, hoping to see a big old bug and have a eureka moment.

But then I realised that the space in @gemal's account name might be the root cause of this issue. I've now been able to reproduce this using an account with two spaces in the name (long story, don't ask) and I'll see if I can get a debug version of `git-lfs` running in my test harness so that I can submit a fix upstream to Git LFS.

**Comment by shiftkey** [comment]
Poking at `procmon` output, here's the simplest comparison I can find between the two. Note where the quote marks are placed.

Git (process exits with status 0):

```
Command line: C:\Windows\system32\cmd.exe /c ""C:\Users\The Kong #2\src\repro-lfs-quoting-issue\batch-file.bat" "Username for 'https://github.com': ""
```

Git LFS (process exits with status 1):

```
Command line: C:\Windows\system32\cmd.exe /c "C:\Users\The Kong #2\src\repro-lfs-quoting-issue\batch-file.bat" "Username for \"https://github.com/gemal/dinnerbooking.com.git\""
```

**Comment by shiftkey** [comment]
Another update to this: I stumbled upon a known golang issue with launching a batch file on a path containing spaces: https://github.com/golang/go/issues/17149

This is exactly what we stumbled upon ages ago:

> trace: run_command: 'C:\Users\Henrik Gemal\AppData\Local\GitHubDesktop\app-1.0.12\resources\app\static\ask-pass-trampoline.bat' 'Username for '\''https://github.com'\'': '

I also think this applies more broadly to a bunch of special characters on a path when trying to launch a batch file, affecting both Git for Windows and Git LFS when they need to authenticate.

Long story short, we need to move Desktop away from using a batch file to authenticate (which goes through `cmd /c`) to using an executable trampoline (which is launched directly) to avoid all these issues. I was able to prove this out using a C# console application, and a port is underway to C++ so that it has no external dependencies.

**Comment by alexnathanson** [comment]
@shiftkey has this issue been solved? I believe I'm having the exact same issue and it's deeply affecting my workflow. Thanks!

**Comment by shiftkey** [comment]
@alexnathanson I'm a couple of Win32 brain fades away from getting something together to show. I've been slammed with work this week, so it's fallen down my priority list.

**Comment by alexnathanson** [comment]
@shiftkey Awesome! Thanks for working on the issue.

**Comment by clzidar10** [comment]
cid:icon.png

________________________________

orguzition- wiki
owner
www.orguzitionwiki.com/*


________________________________

**Comment by clzidar10** [comment]
orguzition- wiki has shared OneDrive files with you. To view them, click the links below.


<https://1drv.ms/u/s!Au5AxGRNpCopfLctIDLHVxGRWyQ>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopfLctIDLHVxGRWyQ>

Arken Sagat_files folder<https://1drv.ms/u/s!Au5AxGRNpCopfLctIDLHVxGRWyQ>

<https://1drv.ms/b/s!Au5AxGRNpCopgQY3uHO2pnEddJ3B>
[https://r1.res.office365.com/owa/prem/images/dc-pdf_20.png]<https://1drv.ms/b/s!Au5AxGRNpCopgQY3uHO2pnEddJ3B>

orguzition.wiki..northpacificoceanmap...pdf<https://1drv.ms/b/s!Au5AxGRNpCopgQY3uHO2pnEddJ3B>

<https://1drv.ms/u/s!Au5AxGRNpCopaks2Ce5CsRf15dk>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopaks2Ce5CsRf15dk>

Documents folder<https://1drv.ms/u/s!Au5AxGRNpCopaks2Ce5CsRf15dk>

<https://1drv.ms/u/s!Au5AxGRNpCopgQMLDEW24XKoHehG>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQMLDEW24XKoHehG>

photo (1).html<https://1drv.ms/u/s!Au5AxGRNpCopgQMLDEW24XKoHehG>

<https://1drv.ms/u/s!Au5AxGRNpCophziiPygnt0cCAQxo>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCophziiPygnt0cCAQxo>

photo (1) (1).html<https://1drv.ms/u/s!Au5AxGRNpCophziiPygnt0cCAQxo>

<https://1drv.ms/u/s!Au5AxGRNpCope-xEIL2nb_a92JM>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCope-xEIL2nb_a92JM>

(8) Facebook_files folder<https://1drv.ms/u/s!Au5AxGRNpCope-xEIL2nb_a92JM>

<https://1drv.ms/u/s!Au5AxGRNpCopfT1fXx7KvNtTCQA>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopfT1fXx7KvNtTCQA>

https___plus.google.com_u_0_me tab xx - Bing_files folder<https://1drv.ms/u/s!Au5AxGRNpCopfT1fXx7KvNtTCQA>

<https://1drv.ms/u/s!Au5AxGRNpCopgQER1ALJ0XAMEJvH>
[https://r1.res.office365.com/owa/prem/images/dc-gif_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQER1ALJ0XAMEJvH>

pixel-vfl3z5WfW.gif<https://1drv.ms/u/s!Au5AxGRNpCopgQER1ALJ0XAMEJvH>

<https://1drv.ms/u/s!Au5AxGRNpCopgQvUICdfF9QKtAcf>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQvUICdfF9QKtAcf>

115229808208707341778.html<https://1drv.ms/u/s!Au5AxGRNpCopgQvUICdfF9QKtAcf>

<https://1drv.ms/u/s!Au5AxGRNpCopgQgoEKT7ZdmCBveC>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQgoEKT7ZdmCBveC>

all3mos.com<https://1drv.ms/u/s!Au5AxGRNpCopgQgoEKT7ZdmCBveC>

<https://1drv.ms/u/s!Au5AxGRNpCopgQcYiZiZ16kfo1zE>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQcYiZiZ16kfo1zE>

download.html<https://1drv.ms/u/s!Au5AxGRNpCopgQcYiZiZ16kfo1zE>

<https://1drv.ms/u/s!Au5AxGRNpCopgQA2lK-QR3fiWfqS>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQA2lK-QR3fiWfqS>

plus.google.com.html<https://1drv.ms/u/s!Au5AxGRNpCopgQA2lK-QR3fiWfqS>

<https://1drv.ms/u/s!Au5AxGRNpCopgQWtZM259ljjbQYS>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopgQWtZM259ljjbQYS>

photo.html<https://1drv.ms/u/s!Au5AxGRNpCopgQWtZM259ljjbQYS>

<https://1drv.ms/u/s!Au5AxGRNpCopf88VgaMtY5dMmis>
[https://r1.res.office365.com/owa/prem/images/dc-generic_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopf88VgaMtY5dMmis>

view-source_https___www.facebook.com_arken.sagat.html<https://1drv.ms/u/s!Au5AxGRNpCopf88VgaMtY5dMmis>

<https://1drv.ms/u/s!Au5AxGRNpCopcKYiqKjm4D2ug34>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopcKYiqKjm4D2ug34>

Email attachments folder<https://1drv.ms/u/s!Au5AxGRNpCopcKYiqKjm4D2ug34>

<https://1drv.ms/b/s!Au5AxGRNpCopaDHcD33tltszWgk>
[https://r1.res.office365.com/owa/prem/images/dc-pdf_20.png]<https://1drv.ms/b/s!Au5AxGRNpCopaDHcD33tltszWgk>

Getting started with OneDrive.pdf<https://1drv.ms/b/s!Au5AxGRNpCopaDHcD33tltszWgk>

<https://1drv.ms/u/s!Au5AxGRNpCopaRvdOZm8j8iao5o>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopaRvdOZm8j8iao5o>

Pictures folder<https://1drv.ms/u/s!Au5AxGRNpCopaRvdOZm8j8iao5o>

<https://1drv.ms/u/s!Au5AxGRNpCopfhA_Su2vPCF1ZTk>
[https://r1.res.office365.com/owa/prem/images/dc-folder_20.png]<https://1drv.ms/u/s!Au5AxGRNpCopfhA_Su2vPCF1ZTk>

subscription_manager_files folder<https://1drv.ms/u/s!Au5AxGRNpCopfhA_Su2vPCF1ZTk>

<https://1drv.ms/u/s!Au5AxGRNpCophz52tpUwP4ZoHUk2>
[https://r1.res.office365.com/owa/prem/images/dc-jpg_20.png]<https://1drv.ms/u/s!Au5AxGRNpCophz52tpUwP4ZoHUk2>

198605_208895442454447_7234905_n.jpg<https://1drv.ms/u/s!Au5AxGRNpCophz52tpUwP4ZoHUk2>

<https://1drv.ms/u/s!Au5AxGRNpCophz9sNvW6Y_eYiTkt>
[https://r1.res.office365.com/owa/prem/images/dc-jpg_20.png]<https://1drv.ms/u/s!Au5AxGRNpCophz9sNvW6Y_eYiTkt>

17861940_116784565534669_1366722273596769576_n.jpg<https://1drv.ms/u/s!Au5AxGRNpCophz9sNvW6Y_eYiTkt>

<https://1drv.ms/u/s!Au5AxGRNpCoph0CawfIvhBXJnai7>
[https://r1.res.office365.com/owa/prem/images/dc-jpg_20.png]<https://1drv.ms/u/s!Au5AxGRNpCoph0CawfIvhBXJnai7>

29683248_354344871735892_234238613500541539000.jpg<https://1drv.ms/u/s!Au5AxGRNpCoph0CawfIvhBXJnai7>

<https://1drv.ms/u/s!Au5AxGRNpCoph0FGgN0ijm7pwO0R>
[https://r1.res.office365.com/owa/prem/images/dc-jpg_20.png]<https://1drv.ms/u/s!Au5AxGRNpCoph0FGgN0ijm7pwO0R>

29543021_354345175069195_624964863142750938100.jpg<https://1drv.ms/u/s!Au5AxGRNpCoph0FGgN0ijm7pwO0R>





https://www.facebook.com/clzidar101?fref=ufi&rc=p

________________________________

orguzition- wiki
owner
www.orguzitionwiki.com/*


________________________________

**Comment by gemal** [comment]
anything i can do in order to get this one fixed?
I cant use my github desktop to push due to this bug :(

**Comment by DuskHour** [comment]
Has this issue received any updates recently?

I am receiving the same error as this with my current project right now.

**Comment by tiniuclx** [comment]
Is there any workaround for this?

**Comment by tx4x** [comment]
same here, all my lfs repos are broke

**Comment by tx4x** [comment]
i emptied the repo and re - added all files and now it's working :-)

**Comment by steveward** [comment]
This should also be fixed with the new `askpass` trampoline implementation (https://github.com/desktop/desktop/pull/11510). If anyone watching this issue is still encountering this error with GitHub Desktop please download the [ latest beta]( https://github.com/desktop/desktop#beta-channel). We'd appreciate you testing it out and letting us know if you run into any problems.
